lib/matplotlib/bezier.py

```diff
@@ -177,18 +225,70 @@ class BezierSegment:
     """
 
     def __init__(self, control_points):
-        n = len(control_points)
-        self._orders = np.arange(n)
-        coeff = [math.factorial(n - 1)
-                 // (math.factorial(i) * math.factorial(n - 1 - i))
-                 for i in range(n)]
-        self._px = np.asarray(control_points).T * coeff
+        self.cpoints = np.asarray(control_points)
+        self.n, self.d = self.cpoints.shape
+        self._orders = np.arange(self.n)
+        coeff = [math.factorial(self.n - 1)
+                 // (math.factorial(i) * math.factorial(self.n - 1 - i))
+                 for i in range(self.n)]
+        self._px = self.cpoints.T * coeff
 
     def point_at_t(self, t):
         """Return the point on the Bezier curve for parameter *t*."""
         return tuple(
             self._px @ (((1 - t) ** self._orders)[::-1] * t ** self._orders))
 
+    def __call__(self, t):
+        return self.point_at_t(t)
+
+    @property
+    def interior_extrema(self):
+        """
+        Return the location along the curve's interior where its partial
+        derivative is zero, along with the dimension along which it is zero for
+        each instance.
+
+        Returns
+        -------
+        dims : int, array_like
+            dimension $i$ along which the corresponding zero occurs
+        dzeros : float, array_like
+            of same size as dims. the $t$ such that $d/dx_i B(t) = 0$
+        """
+        if self.n <= 2:  # a line's extrema are always its tips
+            p = [0]  # roots returns empty array for zero polynomial
+            return np.array([]), np.array([])
+        elif self.n == 3:  # quadratic curve
+            # the bezier curve in standard form is
+            # cp[0] * (1 - t)^2 + cp[1] * 2t(1-t) + cp[2] * t^2
+            # can be re-written as
+            # cp[0] + 2 (cp[1] - cp[0]) t + (cp[2] - 2 cp[1] + cp[0]) t^2
+            # which has simple derivative
+            # 2*(cp[2] - 2*cp[1] + cp[0]) t + 2*(cp[1] - cp[0])
+            p = [2*(self.cpoints[2] - 2*self.cpoints[1] + self.cpoints[0]),
+                 2*(self.cpoints[1] - self.cpoints[0])]
+        elif self.n == 4:  # cubic curve
+            P = self.cpoints
+            # derivative of cubic bezier curve has coefficients
+            a = 3*(P[3] - 3*P[2] + 3*P[1] - P[0])
+            b = 6*(P[2] - 2*P[1] + P[0])
+            c = 3*(P[1] - P[0])
+            p = [a, b, c]
+        else:  # self.n > 4:
+            # a general formula exists, but is not useful for matplotlib
```

Comment:
The general formula (the derivative of https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Polynomial_form) doesn't seem too bad to type in?

Reply:
I am happy to type in the general formula (which I think is harder to read than what I currently have, to be honest), but even if we did so, we'd still probably want to throw NotImplementedError or at least a Warning for high-order bezier curves, since the formulas you linked to have pretty bad numerical properties as order increases.

Reply:
plus, it contains binomial coefficients, so I'd have to add another helper method to compute those, since we don't already depend on scipy.

Reply:
A warning seems fine?

Reply:
Added warning, and implemented math.comb myself to prevent requiring Python>=3.8.